Python/compile.c

```diff
@@ -639,6 +704,15 @@ compiler_setup(struct compiler *c, mod_ty mod, PyObject *filename,
     }
 
     c->c_filename = Py_NewRef(filename);
+    const char *f = PyUnicode_AsUTF8(c->c_filename);
+    if (f == NULL) {
+        PyErr_Clear();
+        c->c_regcode = false;
+    }
+    else {
+        c->c_regcode = strstr(f, "mytest");
```

Comment:
It seems to work now when I run a small test file, but if I set c_regcode to always be true then it seems to pick up some stale frozen bytecode or something like that.

Reply:
./_bootstrap_python ./Programs/_freeze_module.py abc ./Lib/abc.py Python/frozen_modules/abc.h
[1138](https://github.com/python/cpython/actions/runs/3721027867/jobs/6310902255#step:10:1139)
_bootstrap_python: Python/generated_cases.c.h:52: _PyEval_EvalFrameDefault: Assertion `value != NULL' failed.

Reply:
Did I mess up something in marshal?

Reply:
Fix a bug, and now I get a different error:

During 
`./_bootstrap_python ./Programs/_freeze_module.py abc ./Lib/abc.py Python/frozen_modules/abc.h`

it calls PyFunction_NewWithQualName and code_obj is freed:

```
frame #4: 0x0000000100178715 _bootstrap_python`_PyEval_EvalFrameDefault(tstate=<unavailable>, frame=<unavailable>, throwflag=0) at generated_cases.c.h:3731:17 [opt]
   3728	        TARGET(MAKE_FUNCTION) {
   3729	            PyObject *codeobj = POP();
   3730	            PyFunctionObject *func = (PyFunctionObject *)
-> 3731	                PyFunction_New(codeobj, GLOBALS());
   3732	
   3733	            Py_DECREF(codeobj);
   3734	            if (func == NULL) {
(lldb) p _PyObject_Dump((PyObject*)codeobj)
<object at 0x101274240 is freed>
```

(This is why I'm wondering about marshal.)

Reply:
I don't think it's marshal. I set `c_regcode = true` and got the same crash. But before it runs `_bootstrap_python` (which crashes) it runs
```
./Programs/_freeze_module importlib._bootstrap ./Lib/importlib/_bootstrap.py Python/frozen_modules/importlib._bootstrap.h
```
(and two similar ones), which doesn't crash, and the output file is the same as with `c_regcode = false`.

Reply:
I found an easier way to debug, at least bypassing the _bootstrap_python crashes:
```
        c->c_regcode = !strstr(f, "import");
```
This basically bypasses the register VM for importlib.

Now I can run various "real" things and get it to crash in places where it's a little easier to debug (maybe).

I wonder if the culprit might be 32b35d0e6a "put the tmp registers after the stack"?

Reply:
But I see nothing wrong with that commit, nor with anything else. All I see is that some register points to freed memory (filled with 0xdddddddd) instead of being either NULL or a valid object.

Reply:
Maybe we should invest in some better infrastructure to debug these kinds of things. E.g. speed up the build by not deep-freezing anything (freeze only the importlib bootstrap modules), and ignore `__pycache__` directories, so you don't have to worry about the magic number. And upon a crash, somehow dump out the instructions around the crash, registers and stack, etc.

Reply:
I see a problem in take_ownership (Python/frame.c) - it doesn't copy over the the tmps, only copies up to stacktop. Should do the same as in _PyFrame_Copy which is just above it.

We also need to traverse the tmps in _PyFrame_Clear and _PyFrame_Traverse (but not doing that should cause refleaks rather than crashes).  And there are also some functions in Objects/frameobject.c that need to be updated for the tmps. Trying to figure it all out.

Reply:
Making everything at runtime believe that the temps are just part of the locals would seem simpler?

Reply:
That was my first approach, but it got complicated around the places where we deduplicate the varnames which are also cellvars, and then everything gets shifted. (That bit of logic shows up in  fix_cell_offsets, in compute_localsplus_info and PyCode_NewWithPosOnlyArgs.) So I waited until I knew what nlocalsplus is, and allocated the tmps just before the stack. But I still had a problem with the MAKE_CELL instructions overriding them. So I gave up and went for this.

Reply:
I could actually add them as locals with name "$I", but then these names would be in the code objects co_names (right?) which seems like a waste.

Reply:
Seems useful for debugging actually.